File: src/main/java/org/example/kalkulationsprogramm/controller/AnalyticsSnapshotIngressController.java
package org.example.kalkulationsprogramm.controller;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.validation.ConstraintViolation;
import jakarta.validation.Validator;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.example.kalkulationsprogramm.domain.WebsiteAnalyticsSnapshot;
import org.example.kalkulationsprogramm.dto.WebsiteAnalytics.AnalyticsSnapshotRequestDto;
import org.example.kalkulationsprogramm.service.WebsiteAnalyticsSnapshotService;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * Public S2S-Eingang fuer den taeglichen Analytics-Snapshot der Marketing-
 * Webseite (bauschlosserei-kuhn.de).
 * <p>
 * Auth: laeuft unter der {@code /api/internal/**}-Filterchain in
 * {@link org.example.kalkulationsprogramm.config.SecurityConfig#funnelFilterChain}
 * - Tailscale-VPN als Transportschicht plus optionaler Cloudflare-Access-JWT
 * via {@link org.example.kalkulationsprogramm.config.CloudflareAccessJwtFilter}
 * (Property {@code cloudflare.access.enabled=true} schaltet die Signaturpruefung
 * scharf). VPN allein zaehlt nicht als Auth - Production muss
 * {@code cloudflare.access.enabled=true} setzen.
 * <p>
 * Spezifikation des Payloads liegt im Website-Repo
 * (docs ANALYTICS_SNAPSHOT_API.md, schemaVersion=1).
 * <p>
 * Wir lesen den Body bewusst als rohen JSON-String ein und parsen erst danach
 * in das DTO. So bleibt {@code raw_payload} byte-identisch zum eingehenden
 * Push - additive Schema-Erweiterungen vom Website-Team gehen nicht verloren.
 */
@Slf4j
@RestController
@RequestMapping("/api/internal/analytics-snapshot")
@RequiredArgsConstructor
public class AnalyticsSnapshotIngressController {

    /**
     * Hartes Body-Limit fuer den eingehenden Snapshot. Realistische Payloads
     * liegen im einstelligen KB-Bereich; alles ueber 1 MB ist entweder kaputt
     * oder ein Missbrauchsversuch.
     */
    static final int MAX_BODY_BYTES = 1_000_000;

    private final WebsiteAnalyticsSnapshotService service;
    private final ObjectMapper objectMapper;
    private final Validator validator;

    @PostMapping(consumes = MediaType.APPLICATION_JSON_VALUE)
    public ResponseEntity<Map<String, Object>> empfangeSnapshot(@RequestBody String rawPayload) {
        if (rawPayload != null && rawPayload.length() > MAX_BODY_BYTES) {
            log.warn("Analytics-Snapshot abgelehnt: Payload zu gross ({} Zeichen).", rawPayload.length());
            return ResponseEntity.status(HttpStatus.PAYLOAD_TOO_LARGE).body(Map.of(
                    "success", false,
                    "message", "Payload zu gross."
            ));
        }

        AnalyticsSnapshotRequestDto dto;
        try {
            dto = objectMapper.readValue(rawPayload, AnalyticsSnapshotRequestDto.class);
        } catch (JsonProcessingException e) {
            log.warn("Analytics-Snapshot Payload ist kein gueltiges JSON: {}", e.getOriginalMessage());
            return ResponseEntity.badRequest().body(Map.of(
                    "success", false,
                    "message", "Payload ist kein gueltiges JSON."
            ));
        }

        Set<ConstraintViolation<AnalyticsSnapshotRequestDto>> violations = validator.validate(dto);
        if (!violations.isEmpty()) {
            String details = violations.stream()
                    .map(v -> v.getPropertyPath() + " " + v.getMessage())
                    .collect(Collectors.joining(", "));
            log.warn("Analytics-Snapshot Validierung fehlgeschlagen: {}", details);
            return ResponseEntity.badRequest().body(Map.of(
                    "success", false,
                    "message", "Validierung fehlgeschlagen.",
                    "details", details
            ));
        }

        try {
            WebsiteAnalyticsSnapshot saved = service.upsert(dto, rawPayload);
            return ResponseEntity.status(HttpStatus.CREATED).body(Map.of(
                    "success", true,
                    "snapshotDate", saved.getSnapshotDate().toString(),
                    "id", saved.getId()
            ));
        } catch (IllegalArgumentException e) {
            log.warn("Analytics-Snapshot abgelehnt: {}", e.getMessage());
            return ResponseEntity.badRequest().body(Map.of(
                    "success", false,
                    "message", e.getMessage()
            ));
        } catch (Exception e) {
            log.error("Fehler beim Speichern des Analytics-Snapshots", e);
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(Map.of(
                    "success", false,
                    "message", "Interner Fehler."
            ));
        }
    }
}

File: src/main/kotlin/org/example/kalkulationsprogramm/controller/AnalyticsSnapshotIngressController.kt
package org.example.kalkulationsprogramm.controller

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import jakarta.validation.Validator
import org.example.kalkulationsprogramm.dto.WebsiteAnalytics.AnalyticsSnapshotRequestDto
import org.example.kalkulationsprogramm.service.WebsiteAnalyticsSnapshotService
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import java.util.stream.Collectors

@RestController
@RequestMapping("/api/internal/analytics-snapshot")
class AnalyticsSnapshotIngressController(
    private val service: WebsiteAnalyticsSnapshotService,
    private val objectMapper: ObjectMapper,
    private val validator: Validator
) {
    @PostMapping(consumes = [MediaType.APPLICATION_JSON_VALUE])
    fun empfangeSnapshot(@RequestBody rawPayload: String?): ResponseEntity<Map<String, Any?>> {
        if (rawPayload != null && rawPayload.length > MAX_BODY_BYTES) {
            log.warn("Analytics-Snapshot abgelehnt: Payload zu gross ({} Zeichen).", rawPayload.length)
            return ResponseEntity.status(HttpStatus.PAYLOAD_TOO_LARGE).body(mapOf("success" to false, "message" to "Payload zu gross."))
        }
        val dto = try {
            objectMapper.readValue(rawPayload, AnalyticsSnapshotRequestDto::class.java)
        } catch (e: JsonProcessingException) {
            log.warn("Analytics-Snapshot Payload ist kein gueltiges JSON: {}", e.originalMessage)
            return ResponseEntity.badRequest().body(mapOf("success" to false, "message" to "Payload ist kein gueltiges JSON."))
        }

        val violations = validator.validate(dto)
        if (violations.isNotEmpty()) {
            val details = violations.stream().map { "${it.propertyPath} ${it.message}" }.collect(Collectors.joining(", "))
            log.warn("Analytics-Snapshot Validierung fehlgeschlagen: {}", details)
            return ResponseEntity.badRequest().body(mapOf("success" to false, "message" to "Validierung fehlgeschlagen.", "details" to details))
        }

        return try {
            val saved = service.upsert(dto, rawPayload)
            ResponseEntity.status(HttpStatus.CREATED).body(
                mapOf("success" to true, "snapshotDate" to saved.snapshotDate.toString(), "id" to saved.id)
            )
        } catch (e: IllegalArgumentException) {
            log.warn("Analytics-Snapshot abgelehnt: {}", e.message)
            ResponseEntity.badRequest().body(mapOf("success" to false, "message" to e.message))
        } catch (e: Exception) {
            log.error("Fehler beim Speichern des Analytics-Snapshots", e)
            ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(mapOf("success" to false, "message" to "Interner Fehler."))
        }
    }

    companion object {
        const val MAX_BODY_BYTES = 1_000_000
        private val log = LoggerFactory.getLogger(AnalyticsSnapshotIngressController::class.java)
    }
}
